Match ingress to service using endpoints

bl1nk opened the pull request:
This commit changes matching of ingresses to services to use the containerPort intead of the targetPort. Using the targetPort can lead to skipper being unable to route traffic to the target.

Two tests in TestConvertPathRule were failing but are not valid cases to test as Kubernetes will not allow you to create a service with two ports of the same name. They have been removed.

Fixes #1645

szuecs commented on dataclients/kubernetes/testdata/ingress/ingress-data/ing-with-prule-named-port.yaml:
```diff
@@ -41,6 +41,6 @@ subsets:
   - ip: 10.2.9.104
   ports:
-  - name: bar
+  - name: baz
```
name does not matter as long targetPort is referenced by number to the ep port

bl1nk replied:
Isn't this a misconception? I understood it the following way: Kubernetes will reference the service port names, not the targert port names in the endpoint ports. 

So let's say we have the following resources on the cluster

```yaml
---
apiVersion: v1
kind: Service
metadata:
  labels:
    app: whoami-deployment
  name: whoami-service
  namespace: default
spec:
  clusterIP: 10.102.22.101
  ports:
  - name: this-is-my-service-port-name
    port: 8080
    protocol: TCP
    targetPort: my-port
  selector:
    app: whoami
  sessionAffinity: None
  type: ClusterIP
---
apiVersion: extensions/v1beta1
kind: Ingress
metadata:
  labels:
    app: whoami
  name: whoami-ingress
  namespace: default
spec:
  rules:
  - host: whoami.test.domain.com
    http:
      paths:
      - backend:
          serviceName: whoami-service
          servicePort: 8080
        pathType: ImplementationSpecific
```

Then the endpoint created by Kubernetes looks like this

```yaml
apiVersion: v1
kind: Endpoints
metadata:
  labels:
    app: whoami-deployment
  name: whoami-service
  namespace: default
subsets:
- addresses:
  - ip: 172.17.0.3
    nodeName: minikube
    targetRef:
      kind: Pod
      name: whoami-deployment-6786bf95fd-fnqnq
      namespace: default
  ports:
  - name: this-is-my-service-port-name
    port: 80
    protocol: TCP
```

Which indeed references the endpoint port by service port name `this-is-my-service-port-name` not the target port name `my-port`.

Unless I am missing something important I think my change is needed to reflect how Kubernetes would create the endpoints in the testcase.

szuecs replied:
thanks reviewing this lead me to write https://github.com/zalando/skipper/issues/1645#issuecomment-742755534

bl1nk replied:
Thanks for the writeup!

szuecs commented on dataclients/kubernetes/ingressdefinitions_test.go:
```diff
@@ -38,16 +65,76 @@ func TestGetTargetPort(t *testing.T) {
 			errExpected: true,
 		},
+		{
+			name: "multiple service ports, by value",
```
please use the new fixture test suite

bl1nk replied:
I added tests covering these cases using the new fixture test suite in f5fda805e43059c699043f9a1d81e6b0f6ccc7cc

This test was testing `getServicePort()` specifically as it replaced `getTargetPort()`. Would you like me to remove these tests and only rely on the new tests instead?

szuecs replied:
yes only the fixture tests

bl1nk replied:
Done. I put the tests into `dataclients/kubernetes/testdata/ingress/service-ports`. Is there any naming convention to follow? Happy to move the tests into `named-ports` or `ingress-data` if you see them fit one of these.

bl1nk commented:
Rebased onto master, Also fixed naming of the endpoints introduced in 4dcbf4bfffaf6d8f31bc7397dbd05181f4ff6857. Please take a look.

szuecs commented:
:+1:

aryszka commented:
Hi @bl1nk , thanks for the PR!

Generally looks good, but I think it still doesn't cover all the cases. This condition checks the endpoints.subsets.ports.name: https://github.com/zalando/skipper/pull/1646/files#diff-bcaf7ab9e470ce06d42cb44a2ac2cce9ef3d31e07e6a7050b81441edb9409141R77

But  in my observation, this name in the endpoints is not always present. I think we can fix this such, that if there is only one port in the endpoints, then use that, otherwise use the matching by name of the current PR.

I'll verify this now in a cluster with a couple of cases.

aryszka commented:
Just to confirm: I checked a few cases, extending this solution to return the single endpoint port when there is only one, should work.

aryszka commented:
would add this to the condition "...when there is only one": **"and the endpoint port has no name"**

szuecs commented:
@bl1nk please fix as @aryszka suggested, I confirmed it in sig-network and we tested it.

bl1nk commented:
Added the condition to return the endpoint of only one endpoint exists. Also added a test that verifies that an endpoint is returned if only one exists and it has no name. Please take a look again!

aryszka commented:
:+1:

szuecs commented on dataclients/kubernetes/testdata/ingress/service-ports/multiple-by-value.yaml:
```diff
@@ -0,0 +1,72 @@
+          name: myapp-deployment-6786bf95fd-fnqnq
+    ports:
+      - name: web
+        port: 80
```
This does not look correct to me. It should be 8080, because it needs to be in sync with pod container spec ports

bl1nk replied:
I agree, fixed and pushed

szuecs commented:
:+1:

szuecs commented:
I think there are some duplicate tests (named ports tests) and some specs that do not exist in Kubernetes. I will clean this up later.

szuecs commented:
@bl1nk that was fast :D thanks !

szuecs commented:
:+1:

aryszka commented:
:+1:
